server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java

```diff
@@ -1003,36 +1004,45 @@ static Settings aggregateIndexSettings(
                 );
                 validateAdditionalSettings(provider, newAdditionalSettings, additionalIndexSettings);
                 additionalIndexSettings.put(newAdditionalSettings);
+                if (provider.overrulesTemplateAndRequestSettings()) {
```

Comment:
This changes to this file look good. However we do need to port the overruling logic also to `MetadataIndexTemplateService#validateIndexTemplateV2(...)` (line 704 to 716) and `TransportSimulateIndexTemplateAction#resolveTemplate(...)` (line 273 to 286). Otherwise  template validation and simulate index template apis don't mimic the behavior of templates are actually applied at index creation time.

Reply:
Good catch, the provider is used in these places indeed.. 

For the former, I'm on the fence. I think it's fine to create composite templates with values that may not agree - templates may be created once and used in many indexes and at different times ( and licenses, potentially). I'd think it's fine to have a `source: synthetic` setting in a template, so long as it get overwritten at index creation.

Taking a look at the latter.

Reply:
Updated `TransportSimulateIndexTemplateAction`

Reply:
Hm thinking more about the latter too.. Wdyt, shall these be updated? I think we should leave templates alone..

Reply:
I do think the overrule logic should also be added to template validation logic. The overrule logic is a change to the index settings provider infrastructure.  An index settings provider could overrule a setting that cause index creation to fail. By updating the template validation logic to also perform the new overrule logic, we can catch these issue when template is created / updated.

Template validation should mimic index creation as close as possible in order to avoid templates from failing index creation because of invalid settings or mappings. This is why index settings providers are part of template validation to begin with.

Reply:
Per offline discussion, we'll keep template validation as-is to avoid hiding settings in the template that may cause problems. We can revisit this in the future as needed.